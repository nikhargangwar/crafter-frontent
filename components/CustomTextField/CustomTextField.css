.custom-textfield-container {
    display: flex;
    align-items: center;
    flex-direction: row !important;
    justify-content: space-between !important;
    gap: 0px 10px;
    position: relative;
    width: 90% !important;
}

.custom-textfield>* {
    margin: 8px 0px 0px 0px !important;
}

.add-textfield-btn {
    position: absolute;
    left: 300px;
    bottom: 245px;
    color: rgb(19, 51, 124) !important;
}

.custom-config-value {
    width: 50% !important;
}

@media screen and (max-width:768px) {
    .add-textfield-btn {
        left: 155px;
        bottom: 270px
    }
}