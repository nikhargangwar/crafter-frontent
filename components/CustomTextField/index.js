import * as React from 'react';
import './CustomTextField.css';
import Box from '@mui/material/Box';
import TextField from '@mui/material/TextField';
import PropTypes from 'prop-types';

export default function CustomTextField({Key,customConfig,handleCustomConfig}){

  let [newConfigKey , setNewConfigKey] = React.useState('');
  const [newConfigValue , setNewConfigValue] = React.useState('');
  const [keyExists, setKeyExists] = React.useState(false);

  React.useEffect(()=>{
    setNewConfigKey(Key);
    setNewConfigValue(customConfig[Key]||'');
  },[Key,customConfig]);

  React.useEffect(() => {
    const timer = setTimeout(() => {
      if(!keyExists)
      {
        handleCustomConfig(Key,newConfigKey,newConfigValue);
      }
    }, 100);
    return () => {clearTimeout(timer);
    };
  }, [newConfigKey,newConfigValue]);
  
  const handleKeyChange = (event)=>{
    newConfigKey= event.target.value;
    if(Key!==newConfigKey && customConfig[newConfigKey]!==undefined)
    {
      setKeyExists(true);
    }
    else{
      setKeyExists(false);
    }
    setNewConfigKey(event.target.value);
  };
  return(
    <>
      <Box

        sx={{ input: { color: 'white' ,borderRadius:'8px',helperText:{color:'white'}},fieldset:{border:'1px solid #18CCCC'} ,label:{color:'white'}} }
        noValidate
        autoComplete="off"
        className='flex items-center justify-between gap-2 relative w-[90%] my-2'
        style = {
          { fontFamily :  'Open Sans'}
        }
      >
        <TextField className="custom-config-value" id="outlined-basic" label = "Enter Key" variant="outlined" data-testid = "newConfig-key" 
          value={newConfigKey} onChange={handleKeyChange}  size = 'small' error={keyExists} helperText={keyExists ? 'This key already exists' : ''}/>
        <TextField className="custom-config-value" id="outlined-basic" label = "Enter Value" data-testid = "newConfig-value" value={newConfigValue} variant="outlined" size='small' onChange={(event)=>{setNewConfigValue(event.target.value);}}/>
      </Box>
    </>
  );
}

CustomTextField.propTypes = {
  Key: PropTypes.string,
  customConfig: PropTypes.object,
  handleCustomConfig: PropTypes.func,
};