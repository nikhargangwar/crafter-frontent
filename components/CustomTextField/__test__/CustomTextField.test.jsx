import {fireEvent, render ,screen ,act} from '@testing-library/react';
import CustomTextField from '..';
import * as React from 'react';

describe('Frontend Configuration',()=>{

  let mock = {
    Key: '0',
    customConfig: {'':''},
    handleCustomConfig:jest.fn(),
  };

  it('should show empty textfield  when component is rendered',()=>{
    expect(screen.queryByTestId('newConfig-key')).toBeFalsy();
    render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleCustomConfig}/>);
    expect(screen.queryByTestId('newConfig-key')).toBeTruthy();
    expect(screen.queryByTestId('newConfig-key').value).toEqual(undefined);
  });

  it('should update newConfigKey when textfield is updated',()=>{
    expect(screen.queryByTestId('newConfig-key')).toBeFalsy();
    render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleCustomConfig}/>);
    expect(screen.queryByTestId('newConfig-key')).toBeTruthy();
    expect(screen.queryByLabelText('Enter Key')).toBeTruthy();
    const configKey = screen.getByTestId('newConfig-key').querySelector('input');
    fireEvent.change(configKey,{target:{value:'name'}});
    expect(screen.queryByDisplayValue('name')).toBeTruthy();
  });

  it('should update newConfigvalue when textfield is updated',()=>{
    expect(screen.queryByTestId('newConfig-value')).toBeFalsy();
    render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleCustomConfig}/>);
    expect(screen.queryByTestId('newConfig-value')).toBeTruthy();
    const configKey = screen.getByTestId('newConfig-value').querySelector('input');
    fireEvent.change(configKey,{target:{value:'name'}});
    expect(screen.queryByDisplayValue('name')).toBeTruthy();
  });

  it('should update customconfig object by calling event passed as prop when textfield is updated',()=>{
    jest.useFakeTimers();
    expect(screen.queryByTestId('newConfig-key')).toBeFalsy();
    render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleCustomConfig}/>);
    expect(screen.queryByTestId('newConfig-key')).toBeTruthy();
    const configKey = screen.getByTestId('newConfig-key').querySelector('input');
    expect(mock.handleCustomConfig).not.toHaveBeenCalled();
    act(() => {
      fireEvent.change(configKey,{target:{value:'name'}});
    });
    jest.advanceTimersByTime(1000);
    expect(mock.handleCustomConfig).toHaveBeenCalledTimes(1);
  });

  it('should display key already exists error when key already exists',()=>{
    jest.useFakeTimers();
    mock = {
      Key: '1',
      customConfig: {
        'name':'shashwath',
        '':'',
      },
      handleCustomConfig:jest.fn(),
    };
    expect(screen.queryByTestId('newConfig-key')).toBeFalsy();
    render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleCustomConfig}/>);
    expect(screen.queryByTestId('newConfig-key')).toBeTruthy();
    const configKey = screen.getByTestId('newConfig-key').querySelector('input');
    fireEvent.change(configKey,{target:{value:'name'}});
    expect(screen.queryByTestId('newConfig-key').querySelector('p')).toBeTruthy();
    expect(screen.queryByTestId('newConfig-key').querySelector('p').textContent).toEqual('This key already exists');
    jest.advanceTimersByTime(1000);
    expect(mock.handleCustomConfig).toHaveBeenCalledTimes(0);
  });
  it('should render correctly and create a snapshot', () => {
    const {asFragment } =  render(<CustomTextField Key={mock.Key} customConfig={mock.customConfig} handleCustomConfig={mock.handleUserConfig}/>);
    expect(asFragment()).toMatchSnapshot();
  });
});