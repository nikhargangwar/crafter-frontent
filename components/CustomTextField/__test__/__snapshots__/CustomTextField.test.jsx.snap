// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frontend Configuration should render correctly and create a snapshot 1`] = `
<DocumentFragment>
  <div
    autocomplete="off"
    class="flex items-center justify-between gap-2 relative w-[90%] my-2 MuiBox-root css-l36gw3"
    novalidate=""
    style="font-family: Open Sans;"
  >
    <div
      class="MuiFormControl-root MuiTextField-root custom-config-value css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="newConfig-key"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Enter Key
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-basic"
          type="text"
          value="1"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Enter Key
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root custom-config-value css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      data-testid="newConfig-value"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="outlined-basic"
        id="outlined-basic-label"
      >
        Enter Value
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
          id="outlined-basic"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-yjsfm1"
          >
            <span>
              Enter Value
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
